CMakeLists.txt

```diff
@@ -431,12 +431,47 @@ endif()
 
 if(DEFINED BUILD_VERSION)
   set(build_version_argument "-DBUILD_VERSION=${BUILD_VERSION}")
-elseif(EXISTS ${ZEPHYR_BASE}/.git/index)
-  set(git_dependency ${ZEPHYR_BASE}/.git/index)
-else()
-  message(WARNING "ZEPHYR_BASE=${ZEPHYR_BASE} doesn't appear to be a git "
-    "repository, please specify '-DBUILD_VERSION=<version>'")
+elseif(NOT ZEPHYR_GIT_INDEX)
+  if(EXISTS ${ZEPHYR_BASE}/.git/index)
+    set(ZEPHYR_GIT_DIR ${ZEPHYR_BASE}/.git/index CACHE PATH
+      "Path to Zephyr git repository index file" FORCE)
+  elseif(EXISTS ${ZEPHYR_BASE}/.git)
+    # Likely a git-submodule. Let's ask git where the real database is located.
+    find_package(Git QUIET)
+    if(GIT_FOUND)
+      execute_process(
+        COMMAND ${GIT_EXECUTABLE} rev-parse --absolute-git-dir
+        WORKING_DIRECTORY ${ZEPHYR_BASE}
+        OUTPUT_VARIABLE zephyr_git_dir
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+        ERROR_STRIP_TRAILING_WHITESPACE
+        ERROR_VARIABLE stderr
+        RESULT_VARIABLE return_code)
+      if(return_code)
+        message(WARNING "BUILD_VERSION: git rev-parse failed: ${stderr}")
```

Comment:
As we have checked for git repo and failed, then we should also set `ZEPHYR_GIT_INDEX` to `ZEPHYR_GIT_INDEX-NOTFOUND` in the cache.

The same goes for this part:
https://github.com/zephyrproject-rtos/zephyr/blob/ad0ebab42101e93004e09ddcf9f396f534d1eda4/CMakeLists.txt#L465-L468

To avoid duplicated sets, we can do like this:
```
  elseif(NOT ZEPHYR_GIT_INDEX)
    # Assume git index is not found, until proven otherwise
    set(ZEPHYR_GIT_INDEX ZEPHYR_GIT_INDEX-NOTFOUND CACHE PATH
      "Path to Zephyr git repository index file")
    if(EXISTS ${ZEPHYR_BASE}/.git/index)
      ... <rest of code as now> ...
```

This is safe, because the other `set(ZEPHYR_GIT_INDEX ... CACHE ... FORCE)` uses force, so they will overwrite, whereas this first line will only set `-NOTFOUND` if the cache if not already present there.

Remember to remove 
https://github.com/zephyrproject-rtos/zephyr/blob/ad0ebab42101e93004e09ddcf9f396f534d1eda4/CMakeLists.txt#L460-L461
as that is then no longer needed.

Reply:
Makes sense. Should be like that now.